umbraco/Umbraco-CMS
# New Backoffice: Refactor migrations to allow for unscoped migrations

> **Note**
> This PR contains breaking changes, I've created a PR here: #13658 that obsoletes these things.

This PR refactors the current migration system to support unscoped migrations. 

The need for this change arose with the introduction of SQLite since SQLite has some limitations when changing the schema, mainly that you cannot alter columns, or add columns not-null columns with complex data types (such as GUIDS). 

The recommended workaround for this is to add a new table, migrate the data, and delete the old one, however, this causes issues with foreign keys. The recommended approach is to disable foreign keys entirely when doing the migration. However, foreign keys must be disabled outside a transaction, this was impossible with our previous migration system since it always ran all migrations in a single transaction meaning that the transaction will always have been started when your migration is run, so you can't disable FKs.

So the fix for this is to allow for migrations with no outer scope, however, this wouldn't work with our existing pattern of wrapping all migrations in a single scope and only completing this if the entire migration (or migrations for packages) succeeded.

Instead, now every migration step runs in its own scope, or no scope if it inherits from `UnscopedMigrationBase`, this also means that a partial success of a migration is possible for instance this scenario is entirely possible 

* Migration A runs successfully
* Migration B fails 
* Migration C isn't run because B failed

In this case, the changes from migration A are committed and saved, however, migration B is rolled back and C is never rolled. 
Now if you then resolve the issues with migration B and run the upgrader again we'll continue from migration B, since A was successfully completed. 

## Post migrations

Post migrations are no longer a thing, the reasoning for this is partly that it posed the question, should the post migrations run on a partial success? Well, this would very much depend on the type of post-migration, and the scenario. Additionally this was a bit messy with both the entire plan and the individual migration being able to add to this.

Another thing to note with these was that at least in our own code base, these was only every used to rebuild the cache. So to still support this feature the `MigrationBase` now has a flag `RebuildCache` if this is set to true, the cache will be rebuilt if the migration succeeds, although only after the entire plan has run.

As a substitute for the previous migration plan added post migrations, you can now instead use the relevant notifications `UmbracoPlanExecutedNotification` for the internal migrations, and `MigrationPlansExecutedNotification` this notification has all the relevant data about the migration run to allow you to determine what to do, for instance whether the plan succeeded or not, what steps were run and so on, see `ClearCsrfCookieHandler` added in this PR for an example. 


# Testing

I've added some integration tests that cover the new behavior. 

For manual testing you can do the following to test it: 

* Add a test migration that will throw an exception (see below)
* Add a `NoopMigration` to the UmbracoPlan
* Add the `ErrorMigration` after
* Finally add a `NoopMigration`
* Now try and run the upgrader, the upgrade will fail 
* Check the row with the key `Umbraco.Core.Upgrader.State+Umbraco.Core` in the `umbracoKeyValue` table, the value should equal the identifier of your first `NoopMigration` since that's the last one that succeded
* Update the `ErrorMigration` to no longer throw an error
* Run the migration again, it should succeed
* Check the ` umbracoKeyValue` table again, the value should now equal your final `NoopMigration` since it all worked successfully

It might also be a good idea to add a breakpoint into `ClearCsrfCookieHandler` and ensure that the `Handle` method is invoked. 

## Appendix

ErrorMigration:

```C#
namespace Umbraco.Cms.Infrastructure.Migrations;

public class ErrorMigration : MigrationBase
{
    public ErrorMigration(IMigrationContext context) : base(context)
    {
    }

    protected override void Migrate() => throw new NotImplementedException();
}
```

UmbracoPlan example:

```C#
        // Testing
        To<NoopMigration>("{C0B4558A-1793-457F-83ED-F2517D501A84}");
        To<ErrorMigration>("{DC8082BE-3D10-4655-9FD4-4C21A0E9A053}");
        To<NoopMigration>("{9FA75A66-15C7-4332-8832-DB0FB169D834}");
```
